neuro_san/internals/graph/activations/calling_activation.py

```diff
@@ -221,7 +223,16 @@ async def make_one_tool_function_call(self, component_tool_call: ToolCall) -> Di
             self.factory.create_agent_activation(self.run_context, our_agent_spec, use_tool_name,
                                                  self.sly_data, tool_arguments)
 
-        output: str = await callable_component.build()
+        message_list: List[BaseMessage] = await callable_component.build()
+
+        # We got a list of messages back as a string. Take the last.
+        last_message: BaseMessage = message_list[-1]
```

Comment:
Only use the last message of what was returned by build().
This is a bit risky, but if my investigations are telling me what I think they are, this is the only message used at the consumer end anyway.

We are still making a string here (see below), but now at least it can be smaller.